pghoard/basebackup_delta.py

```diff
@@ -0,0 +1,347 @@
+        all_existing_files = self._list_existing_files()
+        for snapshot_file in snapshot_result.state.files:
+            if snapshot_file.hexdigest in all_existing_files:
+                snapshot_file_from_manifest = all_existing_files[snapshot_file.hexdigest]
+                total_stored_size += snapshot_file_from_manifest.stored_file_size
+                total_size += snapshot_file_from_manifest.file_size
+                uploaded_hashes.add(
+                    SnapshotHash(
+                        hexdigest=snapshot_file_from_manifest.hexdigest, size=snapshot_file_from_manifest.file_size
+                    )
+                )
+
+        todo = snapshot_hashes.difference(uploaded_hashes)
+        todo_count = len(todo)
+
+        self.log.info(
+            "Submitting delta hashes to upload: %r, total hashes in the snapshot: %r", todo_count, len(snapshot_hashes)
+        )
+
+        upload_files_count = self._submit_uploading(callback_queue=callback_queue, todo=todo, snapshotter=snapshotter)
+
+        processed_files_count = 0
+        process_futures = []
+        upload_size = 0
+        upload_stored_size = 0
+
+        self.log.info("Waiting for the upload of delta backup files (%r) to finish", upload_files_count)
+
+        def _post_process_delta_file(upload_result):
+            opaque = upload_result["opaque"]
+            storage_delta_path = os.path.join(self.site_config["prefix"], "basebackup_delta")
+            old_key = os.path.join(storage_delta_path, opaque["object_path"])
+            new_key = os.path.join(storage_delta_path, opaque["hexdigest"])
+            self.storage.copy_file(source_key=old_key, destination_key=new_key)
+            self.storage.delete_key(old_key)
+            new_file_size = opaque["new_size"]
+            new_file_stored_size = opaque["result_size"]
+            snapshotter.update_snapshot_file_data(
+                relative_path=opaque["relative_path"],
+                hexdigest=opaque["hexdigest"],
+                file_size=new_file_size,
+                stored_file_size=new_file_stored_size
+            )
+            return new_file_size, new_file_stored_size
+
+        with ThreadPoolExecutor(max_workers=self.parallel) as tpe:
+            while processed_files_count < upload_files_count:
+                try:
+                    delta_upload_result = callback_queue.get(timeout=3.0)
+                    process_futures.append(tpe.submit(_post_process_delta_file, delta_upload_result))
+                    processed_files_count += 1
+                except Empty:
+                    continue
+
+            self.log.info("Waiting for delta backups files post-processing to finish")
+            for future in concurrent.futures.as_completed(process_futures):
+                size, stored_size = future.result()
+                upload_size += size
+                upload_stored_size += stored_size
+
+        self.log.info("All delta file were processed successfully")
```

Comment:
As a general note calling these delta files is a bit confusing. It suggests we keep track of changes within individual files while it's only the full backups themselves that try to reuse full files of other full backups if already available. Not sure if it's worthwhile doing anything about that.

Reply:
Yeah, might be confusing a bit. I think I was thinking too much about `delta` lately :) 
Changed that logging and also in some other places, let me know if there is anything else confusing which I could change.